### PR TITLE
Iterative Explosions

### LINKED ISSUE #2199: [Newmap] Explosives runtime
```
runtime error: 
[11:27:06]Cannot read null.explosion_resistance
[11:27:06]proc name: explosion spread (/datum/controller/subsystem/explosives/proc/explosion_spread)
[11:27:06]  source file: explosives.dm,267
[11:27:06]  usr: null
[11:27:06]  src: Explosives (/datum/controller/subsystem/explosives)
[11:27:06]  call stack:
[11:27:06]Explosives (/datum/controller/subsystem/explosives): explosion spread(null, 1, 8)
[11:27:06]Explosives (/datum/controller/subsystem/explosives): explosion spread(space (1,111,4) (/turf/simulated/open/airless), 1, 8)
[11:27:06]Explosives (/datum/controller/subsystem/explosives): explosion spread(space (2,111,4) (/turf/simulated/open/airless), 1, 8)
[11:27:06]Explosives (/datum/controller/subsystem/explosives): explosion spread(space (3,111,4) (/turf/simulated/open/airless), 1, 8)
[11:27:06]Explosives (/datum/controller/subsystem/explosives): explosion spread(space (4,111,4) (/turf/simulated/open/airless), 1, 8)
[11:27:06]Explosives (/datum/controller/subsystem/explosives): explosion spread(space (5,111,4) (/turf/simulated/open/airless), 1, 8)
[11:27:06]Explosives (/datum/controller/subsystem/explosives): explosion spread(space (6,111,4) (/turf/simulated/open/airless), 1, 8)
[11:27:06]Explosives (/datum/controller/subsystem/explosives): explosion spread(space (7,111,4) (/turf/simulated/open/airless), 1, 1)
[11:27:06]Explosives (/datum/controller/subsystem/explosives): explosion spread(space (7,112,4) (/turf/simulated/open/airless), 1, 4)
[11:27:06]Explosives (/datum/controller/subsystem/explosives): explosion spread(space (8,112,4) (/turf/simulated/open/airless), 1, 2)
[11:27:06]...
[11:27:06]Explosives (/datum/controller/subsystem/explosives): explosion spread(the steel wall (97,129,4) (/turf/simulated/wall), 142, 1)
[11:27:06]Explosives (/datum/controller/subsystem/explosives): explosion spread(the reinforced plasteel wall (97,127,4) (/turf/simulated/wall/r_wall), 173, 1)
[11:27:06]Explosives (/datum/controller/subsystem/explosives): explosion spread(the floor (97,125,4) (/turf/simulated/floor/tiled/dark), 175, 1)
[11:27:06]Explosives (/datum/controller/subsystem/explosives): explosion spread(the floor (97,123,4) (/turf/simulated/floor/tiled), 177, 1)
[11:27:06]Explosives (/datum/controller/subsystem/explosives): explosion spread(the floor (97,121,4) (/turf/simulated/floor/tiled), 179, 1)
[11:27:06]Explosives (/datum/controller/subsystem/explosives): explosion spread(the steel wall (97,119,4) (/turf/simulated/wall), 185, 1)
[11:27:06]Explosives (/datum/controller/subsystem/explosives): explosion spread(the floor (97,117,4) (/turf/simulated/floor/tiled/white), 187, 1)
[11:27:06]Explosives (/datum/controller/subsystem/explosives): explosion spread(the floor (97,115,4) (/turf/simulated/floor/tiled/white), 199, 1)
[11:27:06]Explosives (/datum/controller/subsystem/explosives): fire(0)
[11:27:06]Explosives (/datum/controller/subsystem/explosives): ignite(0)
```

### PR DESCRIPTION
**Pull request by Lohikar**
This PR replaces recursive explosions with a new spreading engine: iterative explosions.
Should behave similarly, but also be more reliable, potentially faster.

Iterative explosions also support simple-explosion-esque directional explosion sounds/shake, and will now traverse Z-levels if the explosion is strong enough.

~~Also changes `playsound()` to transmit sounds across Z-levels. Uses BYOND's 3rd coordinate on `/sound` to make sound sound above/below the player if they're using headphones.~~ Removed for now, can't get Z-falloff working right.

Fixes #2199.

Known issues:
- [x] Openturfs don't attenuate explosions like space turfs do, causing explosions that reach the roof Z to take a long time to run the discovery phase.

**This will require a config update.**

### CODE REVIEW

**Comment by skull132** [comment]
\> 3D Sound

We A E S T H E T I C now.

**Comment by chaoko99** [comment]
Does this work with the Tachyon array?

**Comment by Lohikar** [comment]
No, but recursive explosions never did anyways.

**Comment by chaoko99** [comment]
Toxins really needs that tachyon array.

File: code/controllers/configuration.dm [hunk 1]
```diff
@@ -167,7 +167,7 @@ var/list/gamemode_cache = list()
 
 	var/simultaneous_pm_warning_timeout = 100
 
-	var/use_recursive_explosions = 0 //Defines whether the server uses recursive or circular explosions.
+	var/use_complex_explosions = 0 //Defines whether the server uses iterative or circular explosions.
```
Comment (skull132) [inline code/controllers/configuration.dm]: Consolidation of terms please. The system should either be referred to throughout the work as `complex_explosions` or as `iterative_explosions`. Unless we, for some reason, feel insane enough to have 3 types of explosions, with two of them as "complex".

File: code/controllers/subsystems/explosives.dm [hunk 2]
```diff
@@ -308,5 +424,10 @@ var/datum/controller/subsystem/explosives/SSexplosives
 	var/flash_range
 	var/adminlog
 	var/z_transfer
-	var/is_rec
+	var/complex
```
Comment (skull132) [inline code/controllers/subsystems/explosives.dm]: See above regarding terms.

File: code/controllers/subsystems/explosives.dm [hunk 3]
```diff
@@ -212,78 +216,190 @@ var/datum/controller/subsystem/explosives/SSexplosives
...
+
+		SEARCH_DIR(origin_direction)
+		SEARCH_DIR(turn(origin_direction, 90))
+		SEARCH_DIR(turn(origin_direction, -90))
+
+		CHECK_TICK
+
+	log_debug("iexpl: Discovery completed in [(world.time-time)/10] seconds.")
+	log_debug("iexpl: Beginning SFX phase.")
+	time = world.time
```
Comment (skull132) [inline code/controllers/subsystems/explosives.dm]: What is this magic?

File: code/controllers/subsystems/explosives.dm [hunk 4]
```diff
@@ -212,78 +216,190 @@ var/datum/controller/subsystem/explosives/SSexplosives
...
-		for(var/atom/A in T)
-			A.ex_act(severity)
+		var/severity = 4 - round(max(min( 3, ((act_turfs[T] - T.explosion_resistance) / (max(3,(power/3)))) ) ,1), 1)
+		//sanity			effective power on tile				divided by either 3 or one third the total explosion power
+		//															One third because there are three power levels and I
+		//															want each one to take up a third of the crater
+
+		if (T.simulated)
+			INVOKE_ASYNC(T, /atom/.proc/ex_act, severity)
+			T.ex_act(severity)
```
Comment (Lohikar) [inline code/controllers/subsystems/explosives.dm]: :thinking:

Comment (chaoko99) [reply code/controllers/subsystems/explosives.dm]: Lohikar are you reviewing your own code.

**Pull request merged**